src/expr/src/scalar/func/macros.rs

```diff
@@ -148,35 +151,73 @@ macro_rules! sqlfunc {
             fn $fn_name $($tail)*
         );
     };
-    // Then, make the function fallible if it wasn't
+
+    // Stage 1: Convert function to fallible, if it wasn't already
+
+    // It is important that we check if the return type matches Result<Option<T>>, Result<T>,
+    // Option<T>, and T in that order because once a type is matched by a `ty` typed metavariable
+    // it can no longer be destructured syntactically. So if `$ret_ty:ty` matched
+    // `Result<Option<T>>` as a unit we would no longer be able to infer information about the
+    // structure of the type and therefore wouldn't be able to apply the elision rules.
+    (
+        #[sqlname = $name:expr]
+        fn $fn_name:ident $params:tt -> Result<Option<$ret_ty:ty>, EvalError>
+            $body:block
+    ) => {
+        sqlfunc!(
+            @stage2
```

Comment:
Just for my own edification, is this an idiom in macro matching to "advance" matches along some successive set of linear transformations? Very clever.

Reply:
Basically this is a "hack" you have do to because macros don't encapsulate well. If it was rust here I would be calling something like `sqlfunc_impl` that contains the lower level implementation but with macros this would require that the `sqlfunc_impl!` is also imported in the calling site!

This pattern is described here https://danielkeep.github.io/tlborm/book/pat-internal-rules.html and uses the `@` symbol because it can't collide with anything else. In mind I interpret it as an internal function call